Add sanitizers to CI builds (#5996)
ripple/rippled: BUILD.md

   conan install .. --output-folder . --build missing --settings build_type=Debug
   ```

   If you would like to activate `asan+ubsan`(`Address`) or `tsan+ubsan`(`Thread`) for the build,

>    declare an env. variable as follows and simply use the `sanitizers`

Reviewer: ```suggestion
   declare an environment variable as follows and use the `sanitizers`
```

Author: Done

   conan install .. --output-folder . --build missing --settings build_type=Debug
   ```

   If you would like to activate `asan+ubsan`(`Address`) or `tsan+ubsan`(`Thread`) for the build,
   declare an environment variable as follows and use the `sanitizers`
   profile in the `conan install` command. Make sure you clear your conan cache before doing so.

>    Command: `conan cache clean "*"`

Reviewer: ```suggestion
   profile in the `conan install` command.`
```

If profiles are set up correctly, this a bad advice.
We should aim to fix our ptofiles and make `package_id` be different for different configurations, rather than saying "always remove cache first"

In Clio we don't have to manually delete cache and it all works well (and fast)

Author: I agree. But in case of Conan it seems a bit flaky. I had to clean the cache a-few times to get the builds working properly. Sometimes Conan won't build dependencies. Hence starting with a clean cache seemed much cleaner and reliable.

Reviewer: That probably happened before you introduced `package_id` changes I suggested.
You were using a single word to define configuration change, so that was definitely causing usage of packages which are not expected.
Let's try with the current setup?
Now any change to flags of external dependencies will cause recompilation (as it should), so we should no longer have this problem.

We had exactly the same problem, but since fixing package_id never had it happen again

Author: Yeah, makes sense.

   ```

>    SANITIZERS=Address conan install .. --output-folder . --profile sanitizers --build missing --settings build_type=Debug

Reviewer: ```suggestion
   SANITIZERS=Address conan install .. --output-folder . --profile:all sanitizers --build missing --settings build_type=Debug
```

This is a bug, if you specify `--profile` it will only be host context, but you need all your flags in build context.
For simplicity, we use `--profile:all` everywhere

Author: Done

   conan install .. --output-folder . --build missing --settings build_type=Debug
   ```

   **Sanitizers:** To build dependencies with sanitizer instrumentation, set the

>    `SANITIZERS` environment variable(only once for both conan and cmake) and use the `sanitizers` profile:

Reviewer 2: ```suggestion
   `SANITIZERS` environment variable (for both Conan and CMake) and use the `sanitizers` profile:
```

I don't quite understand what "only once" here refers to.

Also, I would suggest to remove this change and the one you added below for cmake, and to add a new `## Sanitizers` section with the combined content after `## Coverage` and before `## Options`.

In the Options section you should then also remove the `To enable sanitizers, set the `SANITIZERS` environment variable ...` since it's duplicative.

Author: > I don't quite understand what "only once" here refers to.
This is to suggest the dev. to not try to set the env. variable SANITIZERS twice. Once before running conan and next before running cmake, with a potential to accidently or intentionally diff. values.
Example
```bash
export SANITIZERS=Address,UndefinedBehavior
conan...

export SANITIZERS=Address
cmake ...
```

I have made other change as well.

- file named `coverage.`_extension_, with a suitable extension for the report format, or
- directory named `coverage`, with the `index.html` and other files inside, for the `html-details` or `html-nested` report formats.

## Sanitizers

To build dependencies and xrpld with sanitizer instrumentation, set the

> `SANITIZERS` environment variable(only once before running conan and cmake) and use the `sanitizers` profile in conan:

Reviewer: ```suggestion
`SANITIZERS` environment variable (only once before running conan and cmake) and use the `sanitizers` profile in conan:
```
